allow specifying which auto-instrumented urls should have tracing headers

trescenzi opened the pull request:
While the tracing headers are useful when sent to backends that expect, and utilize them, automatic instrumentation is also useful for just tracking the time and response status of requests to external services that aren't being tracked. If such an external service has a strict policy on allowed headers the automatically added headers will cause instrumented requests to error. So instead of not getting any information from those URLs this commit adds the ability to whitelist, and blacklist, the URLs that have tracing headers added to them.

The new options are built off of the existing options and are:

- fetch_url_header_inclusion_patterns
- fetch_url_header_exclusion_patterns
- xhr_url_header_inclusion_patterns
- xhr_url_header_exclusion_patterns

I didn't add any tests because there were not other tests that referenced these automatic instrumentation plugins. Let me know if you think I should add some. I also included the dist, not sure if that's correct or not.

mwear commented on src/plugins/instrument_fetch.js:
```diff
@@ -189,9 +191,11 @@ class InstrumentFetch {
-            Object.keys(headersCarrier).forEach((key) => {
-                if (!request.headers.get(key)) request.headers.set(key, headersCarrier[key]);
-            });
+            if (self._shouldAddHeadersToRequest(tracer, request.url)) {
```
Is it possible to move this up so that we don't have to create `headersCarrier` or make the call to `tracer.inject`?

trescenzi replied:
Makes sense will do.

mwear commented on src/plugins/instrument_xhr.js:
```diff
@@ -293,9 +295,11 @@ class InstrumentXHR {
-            keys.forEach((key) => {
-                proxied.setRequestHeader.call(this, key, headersCarrier[key]);
-            });
+            if (self._shouldAddHeadersToRequest(tracer, this.__tracer_url)) {
```
Same as the previous comment:

> Is it possible to move this up so that we don't have to create `headersCarrier` or make the call to `tracer.inject`?

trescenzi commented:
@mwear would you like me to add tests for the general include/exclude url options as well? Not 100% sure what'll be involved there, gotta look at how tests are written right now, but I imagine they will look very similar and shouldn't be much more to add.

mwear commented:
That would be awesome and a welcome addition @trescenzi!

trescenzi commented:
I've added tests. Not certain this is what you were looking for though. I wasn't sure of a good way to fully test some of this without stubbing out, `fetch` and `XMLHttpRequest` as these are largely testing side effects of global functions. As a result I included mocks for XHR and Fetch into the browser suite. One I npm installed, `fetch-mock`, the other, `FakeXMLHttpRequest`, I copied the dist into `test/browser` as had been done for other libraries used in those tests. It would be easy to copy `fetch-mock`'s dist as well but since it wasn't published to github I went with loading it through node_modules for now.

Let me know it this makes sense or if you had a different approach in mind.

andrewhsu commented:
@obecny can you also have a review of the deps that are brought in?

obecny commented on test/browser/stub-xhr.js:
```diff
@@ -0,0 +1,525 @@
+// SOURCE: https://github.com/pretenderjs/FakeXMLHttpRequest
```
I would strongly advise not to rely on partially extracted code from sinon, we will loose the possibility of easy maintenance. I would rather use the `sinon` as it will be much easier to get any future updates as well as other sinon features if this is going to be needed in future.

trescenzi replied:
Yea if you guys are up for adding sinon then I'll happily add it in. Do you think it should be added as a dev dependency, and sourced from `node_modules`, or the dist copied in like has been done with mocha and chai? Same question for the `fetch-mock` too actually.

My guess is the copying was done initially in in order to prevent breaks from node_modules being reorganized, or distribution changes since inclusion is being done in a script tag.

obecny replied:
For now I would "follow the pattern" so do the same as you added `fetch-mock` unless I'm not aware of something but I don't see a reason to copy the sinon into new folder are you ok with this @mwear ?. I think this template should be refactored anyway to be "dynamic" but this is another issue and not for this PR :)

obecny commented on test/suites/browser/request_imp.js:
```diff
@@ -0,0 +1,217 @@
+            request.respond(200);
+        });
+
+        it("doesnt traces urls that have been excluded", function(done) {
```
`doesn't trace`

obecny commented on test/unittest.html:
```diff
@@ -10,6 +10,11 @@
+        <script src="browser/stub-xhr.js"></script>
+        <script src="../node_modules/fetch-mock/es5/client-bundle.js"></script>
+        <script>
+          fetchMock.mock('*', 200);
```
will this affect all fetch requests in all existing and new tests in future ?, if so I think this should be moved to certain unit test and be called in "beforeEach" block for only desired tests

trescenzi replied:
Good call.

trescenzi replied:
So because of how the plugins grab fetch/xhr on load the mocks need to be setup before the plugin is loaded. I've added much more specific urls, to prevent clashes between tests, and a comment to make it more clear what's going on and why.

obecny replied:
I would argue as I can  see here for example 
http://www.wheresrhys.co.uk/fetch-mock/
where it works exactly this way, it includes the "plugin" then requires fetchMock, then mock it and then reverting afterwards (reset).

trescenzi commented on test/suites/browser/request_imp.js:
```diff
@@ -0,0 +1,217 @@
+            const request = new XMLHttpRequest();
+            request.open("GET", "http://exclude-me.com");
+            request.send();
+            getHeaderKeys().map(key => expect(request.requestHeaders[key]).to.be.a("string"))
```
Both of those make sense. Thanks :+1:

trescenzi commented:
Hey apologies for the delay here, been sidetracked by the world. Hope everyone's doing alright. I'm planning to get to this either tonight or tomorrow night.

trescenzi commented:
I've updated this based on comments and gotten it up to date with master. Tried to make the mocking more clear and moved the header checks into shared function that are also hopefully a little more self documenting. Let me know if there's anything else that needs to be done.

mwear commented:
Sorry for the delay @trescenzi . @obecny and I will review this on Monday.

obecny commented on src/plugins/instrument_fetch.js:
```diff
@@ -252,6 +256,27 @@ class InstrumentFetch {
+        if (opts.fetch_url_header_inclusion_patterns.some((inc) => inc.test(url))) {
+            include = true;
+        }
+        if (opts.fetch_url_header_exclusion_patterns.some((ex) => ex.test(url))) {
```
this `if` statement for `exclusion` could be first condition and then `return false` without variable `include`
then next condition for `inclusion` and `return true` without variable
and as last simply `return false`.
no need for variable and I think there is no need to check for inclusion before exclusion as exclusion is more important

trescenzi replied:
Yea definitively agree. This was copy/pasted from the the way the other exclusion/inclusion logic works. I'll update both since I agree this logic is kinda wonky.

obecny commented on test/unittest.html:
```diff
@@ -10,6 +10,22 @@
+            fetchMock
+              .get(/include-me-header-test/, 200)
+              .get(/exclude-me-header-test/, 200);
+            sinon.useFakeXMLHttpRequest();
```
this should not be here but in unit tests - before starting, together with restoring after the unit tests are finished
for example
https://sinonjs.org/releases/latest/fake-xhr-and-server/

trescenzi replied:
This is here for the same reason the `fetchMock` is. Because once the lightstep code is loaded it immediately grabs window to try and make sure that it's the first thing to get the `XMLHttpRequest` object. If later calls are made to `useFakeXMLHttpRequest`or `fetchMock` they don't do anything because lightstep has wrapped the real objects already.

Do you have any suggestions for workarounds? I considered writing a mock for wrapping these window functions and then either returning the mock, or not, but that requires some knowledge of exactly how these mocks are wrapping the window objects in the first place.

Basically I agree, just given the limitations and how the plugins are written as of today I couldn't come up with something that seemed both effective but also not overkill.

obecny replied:
@trescenzi if this is not possible then @mwear should decide. I think we should create a separate issue to move the fetchMock and sinon into appropriate unit tests. But this is really strange as it looks a bit for me like hacky solution as sinon or fetchMock should be used in a way that you can have full control over the tested class before and after the unit test so you can easily mock it or remove the mock. If this is not possible then something is wrong and should be fixed.

trescenzi replied:
If I'm not mistaken what's happening is that the [lightstep plugin](https://github.com/lightstep/lightstep-tracer-javascript/blob/master/src/plugins/instrument_fetch.js#L6) looks at the window fetch immediately and assigns it as the "proxied" fetch. If in the future something like `fetchMock` mocks fetch it'll grab `window.fetch` and replace it with the mock. However at that point lightstep has already stored its reference to the window's `fetch` which means that even though `fetch` is proxied lightstep will just call through to the old value.

I believe code like the following would work to manage this:
```JavaScript
window._fetch = window.fetch;
Object.define(window, 'fetch', {
  get: function() {
    return this._fetch;
  },
  set: function(newFetch) {
    this._fetch = newFetch;
  }
});
```
So even though lightstep would store a reference to `window.fetch` that reference would be "live" and *should* always call through to the correct, be it the real implementation or the mock one, at the time of usage. I'm going to try this out. It's hacky, but does feel better than just always mocking things.

trescenzi replied:
So with a quick test that makes `fetchMock` not work. I'm really not sure if there's a way to solve this without changing the lightstep code itself. Since it's storing a reference to whatever `window.fetch`/`window.XHMHttpRequest` is at the time it's loaded it makes it really rough to dynamically turn on/off anything that wraps those methods.

trescenzi replied:
So with a quick test that makes `fetchMock` not work. I'm really not sure if there's a way to solve this without changing the lightstep code itself. Since it's storing a reference to whatever `window.fetch`/`window.XHMHttpRequest` is at the time it's loaded it makes it really rough to dynamically turn on/off anything that wraps those methods.

obecny replied:
@trescenzi I understand and I'm fine with keeping it as you did :). But also we should have a new issue to make all what is necessary to be able to use test tools (sinon, fetchmock etc.) correctly inside unit tests, WDYT @mwear ?

trescenzi replied:
Ok cool. I'm largely just continuing to think about this because I agree with you. It's surprising that we can't use these mocks as expected since they are super common, standard solutions for testing network requests.

With regards to an issue I might actually be able to file an issue that's relevant to my usage of the plugins. When incorporating these plugins I actually broke my e2e tests because of a similar problem. I have to go back and look at what happened though to see if it's something that makes sense to be an issue.

mwear replied:
I'm fine with addressing the testing issues separately, as I would really like to get this work in so folks can use it.

mwear commented:
@trescenzi I just wanted to check in on the status of this PR. Is it complete and in as good of a final state as it can be, aside from workarounds that are needed for testing?

trescenzi commented:
@mwear I just pushed the changes that @obecny suggested for the `_shouldTrace`/`_shouldAddHeaders` methods. I separated out the `_shouldTrace` change in case you want to drop it because it's a refactor of existing code. Aside from that I believe it should be good to go.

mwear commented:
@trescenzi This shipped yesterday in v0.30.0 and is available on NPM. Thanks again for all your hard work!
